### DISCUSSION TITLE
Get first image after parsing instead of [img] tag

### DISCUSSION THREAD

**Discussion by Sama34** [RevertIT/mybb-rt_discord_webhooks#2] [Ideas]
MyBB allows for HTML in posts, as well as attachments (which can be visible to guests). Additionally, there are plugins that allow adding images to threads, like xThreads.

I think it would be a nice addition if  instead of getting only the first [img] (if any), the message would be fully parsed and then get the first image by HTML tag.

https://github.com/RevertIT/mybb-rt_discord_webhooks/blob/e06f251a9af5dff28f9d0bdee0bdb1bd5944922f/inc/plugins/rt/src/DiscordWebhooks/Discord/DiscordHelper.php#L82-L93

**Comment by RevertIT** [comment]
While this is a valid suggestion, there are several drawbacks to it.

This happens at insertion time meaning page load speed is crucial, but:
1. Just by calling Discord API, we decrease page load speed;
2. Having to query attachments and search for icons will further affect performance;
3. Then call the parser to parse everything as HTML, then check for the `src` tag will again hit performance;

We could of course do this via task and have 0 performance issues like I did with the ChatGPT plugin, but I don't like the idea of flooding Discord API via scheduled task.

The most we can do is check whether a specific forum has enabled HTML in posts and just search for it in `input['message']`.

If you have some other approach that wouldn't hit performance, you can post it here.

PS: You can test https://github.com/RevertIT/mybb-rt_discord_webhooks/tree/test-image-tags

**Comment by Sama34** [comment]
I suppose an alternative would be to allow this with settings, so only the affected forums under specific setups could get affected when enabling such feature (to get the first image regardless of HTML, BBCode, or other methods).

I didn't dig into the code that much, but would handling Discord Webhooks by a task be possibly a more performance/user experience approach ?

**Comment by RevertIT** [comment]
> I didn't dig into the code that much, but would handling Discord Webhooks by a task be possibly a more performance/user experience approach ?

Yes, but this way it would post a bunch of replies, new threads, and updated posts, all at the same time instead of posting/updating it at posting time. (Which will flood channels with many messages and possibly annoy users of that discord channel)

**Reply by Sama34** [reply]
Ok, I understand, what I meant (to clarify) is that perhaps the task can run every minute or even seconds, but the primary reason would be to move the process outside the scope of the user interaction (posting, registering, etc).

So, say, user won't notice any hang while posting but seconds later the task will run and handle the API on the background.

It should be noted that sites can be configured in such a way that tasks are executed at a server level instead of just loading the `task.php` file with every browser visit.

**Comment by RevertIT** [comment]
As per the original request, I consider this fixed with implementation to get the first image of HTML (when enabled per forum) or BBCode.

However, HTML tags are all stripped out, only bb codes will be rendered on Discord.
